Initial Flutter Guide (#70)
sonr-io/docs: blog/2022-09-07-flutter-guide/index.mdx

---
slug: first-app-in-motor-flutter

> title: Building an App with motor_flutter

Reviewer: ```suggestion
title: Building an App with motor-flutter
```

authors: [prad.snr/, ntindle.snr/]
tags: [flutter, motor, cli]
---

Building apps has never been easier, especially today with tools like Flutter. 
You can build apps in minutes.

<!--{/* TODO: Add flavor to opening *} -->
Let's Build a simple Flutter To Do app.

To start, make sure you have [Flutter installed](https://docs.flutter.dev/get-started/install).
If you're on MacOS, you can run `brew install --cask flutter` to install it.  

:::tip

You can check that Flutter successfully installed with Flutter doctor:

```bash title="terminal"
flutter doctor
```

> If you get any errors, now is a great time to Google™️ for fixes.

Reviewer: ```suggestion
If you run into installation issues, we suggest checking out [stack overflow](https://stackoverflow.com/search?q=flutter+install)!
```

  ```

2. Import the `get` package in our `main.dart` file:

  ```dart title="lib/main.dart"
  import 'package:get/get.dart';
  ```

3. Change the `MyApp` class to return `GetMaterialApp` instead of `MaterialApp`:

  ```dart title="lib/main.dart"
  class MyApp extends StatelessWidget {
    const MyApp({super.key});
  
    // This widget is the root of your application.
    @override
    Widget build(BuildContext context) {
    // This next line is the only change
      return GetMaterialApp(
        title: 'Flutter Demo',
        theme: ThemeData(
          // This is ...
          primarySwatch: Colors.blue,
        ),
        home: const LoginPage(title: "Login here"),
      );
    }
  }
  ```

### Fixing the initialization

1. Change the `main` function to use `Future<void>` instead of `void`:
  > This allows us to use `async` and `await` in the function.

  ```dart title="lib/main.dart"
  Future<void> main() async {
   // ...
    }
  // ...
  }
  ```
2. Initialize the `MotorFlutter` class in the `main` function using `Get`:

  ```dart title="lib/main.dart"
  Future<void> main() async {
    // Don't worry about this line just yet
    WidgetsFlutterBinding.ensureInitialized();

>     // handle the motor intialization response

Reviewer: ```suggestion
    // handle the motor initialization response
```

          'Create Account',
          style: TextStyle(color: Colors.white, fontSize: 18),
          ),
      ),
  ),
  ```

3. Add the following code to the `create_account_page.dart` file.:
  ```dart title="lib/create_account_page.dart"
  import 'package:flutter/material.dart';
  import 'package:get/get.dart';
  import 'package:motor_flutter/motor_flutter.dart';
  import 'package:todo_app/login_page.dart';
  
  class CreateAccountPage extends StatefulWidget {
    const CreateAccountPage({super.key, this.title = "Create Account"});
  
    final String title;
  
    @override
    _CreateAccountPageState createState() => _CreateAccountPageState();
  }
  
  class _CreateAccountPageState extends State<CreateAccountPage> {
    TextEditingController passwordController = TextEditingController();
  
    // This is the function that will be called when the create account is processed on the motor.
    void _handleCreateAccountResponse(CreateAccountResponse? response) {
      if (response != null) {
        print(response);
        // TODO: Store the credentials in the secure storage
        print(response.whoIs);
        Get.off(const LoginPage(title: "Login here"));
      }
    }
  

>     // The meat and potatos of the file is here.

Author: ```suggestion
    // The meat and potatoes of the file are here.
```
